###################################################################################################################
# 													                                                                                      #
# Script to pull CNV data from the TCGA LUAD data files and annotate the genes altered in GRB2 depleted samples.  #
#													                                                                                        #
###################################################################################################################
#													                                                                                        # 
# AUTHOR LIST:												                                                                            #
#	Natalie Stephenson.   										                                                                      #	
#													                                                                                        #
# CURRENT STATUS OF SCRIPT:						                                                                            #
#													                                                                                        #
#	Although survival data has been downloaded and annotated onto one file from this script, the files used in the  # 
# reshaping step performed on MARC1 did not have this information within the script, as it was determined these   #
# could be added at a later time more easily. The latest version of the script is currently at a stage where the  #
# PCA step is being worked out. The point at which errors are occuring, and the errors being given, are annotated # 
# onto the script at the appropriate place.                                                                       #
#													                                                                                        #
# This script currently utilises the Lung Adenocarcinoma dataset from TCGA (LUAD), however, this can be altered   #
# to other cancer types by altering the appropriate sections of the CNV data download step.                       #
#													                                                                                        #
# In addition, this script is currently looking at proteins containing Class I and II PxxP motifs, i.e. the       #
# classical PxxP motifs, the GRB2_SH3.R script creates lists of all eight classes of PxxP motifs, so altering the #           
# list used when reducing the data would offer information on the different PxxP motif containing proteins.       # 
#						                                                                              							          #
###################################################################################################################

# Intro text.
cat("[INSERT PAPER DETAILS HERE] \n\n",
    "This script downloads Lung Adenocarcinoma TCGA copy number variation (CNV) data and processes it to provide a list of gene in each sample with CNV's. \n",
    "NOTE: This script requires files generated by the script GRB2_SH3.R\n",
    sep="")


# Install required packages 
### make sure this is as efficient as possible (i.e. get rid of those not used!)
source("http://bioconductor.org/biocLite.R")
biocLite("TCGAbiolinks")
#biocLite(c("biorMart", "UniProt.ws", "RTCGA", "RTCGA.CNV", "AnnotationDbi", "SummarizedExperiment", "genomeIntervals", "GenomicRanges"))
install.packages("plyr")
install.packages("dplyr")
# install.packages("stringr")
#install.packages("cgdsr")
install.packages("data.table")


# Load required libraries.
### make sure this is as efficient as possible (i.e. get rid of those not used!)
# library(UniProt.ws)
# library(biomaRt)
library(dplyr)
# library(stringr)
# library(cgdsr)
# library(RTCGA)
library(data.table)
library(plyr)
library(TCGAbiolinks)
# library(SummarizedExperiment)
# library(genomeIntervals)
# library(GenomicRanges)
# library(AnnotationDbi)


##################################
##
## DOWNLOADING CNV DATA FROM THE TCGA DATABASE
##
##################################

# Create a GDC query to look into Copy Number Variations within Lung Adenocarcinoma in the TCGA database.
# Reference for creating further queries: https://bioconductor.org/packages/devel/bioc/vignettes/TCGAbiolinks/inst/doc/tcgaBiolinks.html#gdcquery-searching-tcga-open-access-data
# Note: can use data.category = "Transcriptome Profiling"/"Gene Expression" with data.type = "Gene Expression Quantification" [workflow.type ="HTseq - Counts??, flie.type ="normalized_results"/ experimental.strategy = "RNA-Seq"]
# Note: Used "Masked Copy Number Segment" for data.type as they are generated with the same method as "Copy Number Segment" files except that a filtering step is performed that removes Y chromosome and probe sets that were previously indicated to have frequent germline copy-number variation.
# Should I use Legacy datasets???
queryLUAD_CNV <- GDCquery(project = "TCGA-LUAD", 
                      data.category = "Copy Number Variation",
                      data.type = "Masked Copy Number Segment",
                      legacy = FALSE)


# Download the LUAD CNV data
GDCdownload(queryLUAD_CNV)
LUAD_CNV_assay <- GDCprepare(queryLUAD_CNV)


# Sort the CNV data by chromosome position.
LUAD_CNV_sorted <- LUAD_CNV_assay[order(LUAD_CNV_assay$Chromosome, LUAD_CNV_assay$Start, LUAD_CNV_assay$End),]


# Define CNV from Segment Mean (Segment Mean < -0.2 = -1 LOSS, Segment Mean > 0.2 = 1 GAIN, else 0 NORMAL
LUAD_CNV_sorted <- LUAD_CNV_sorted %>%
  mutate(CNV=Segment_Mean)
LUAD_CNV_sorted$CNV <- cut(LUAD_CNV_sorted$CNV, c(-Inf, -0.2, 0.2, Inf), labels = c("-1", "0", "1"), )
write.csv(LUAD_CNV_sorted, "LUAD_CNV_sorted.csv")



##################################
##
## ANNOTATING GENE NAMES ONTO THE CNV DATA
##
##################################

# Annotate genenames onto the CNV data (https://www.biostars.org/p/147916/)

# Import complete gene list generated from the script GRB2_SH3.R.
complete_gene_list <- read.csv(file = "complete_gene_list.csv")
complete_gene_list <- complete_gene_list[order(complete_gene_list$chromosome_name, complete_gene_list$start_position, complete_gene_list$end_position),]

# Reduce the data and assign gene names to the chromosome regions specified.
## NOTE: This step was performed on the ARC3 server details are below:
##                    DATE PERFORMED:     03MAY2017
##                    SUCCESSFUL JOB ID:  14682
##                    QNAME:              24core-128G.q
##                    TIME:               2.3 hours
##                    MEM:                533.6 TBs
##                    MAXVMEME:           73 GB
## NOTE: In order to perform this script on ARC3 the foverlaps utlilsed a LUAD_CNV file that was split into blocks of 5000 rows. The script for this can 
##       be found in the file foverlaps_script_split.R. The resulting data files were recombined using the bind.R script, this was also performed on ARC3,
##       job ID: 18758, 24core-768G.q, 4.3 hours, 3039 TBs and maxvmem 224.7 GB.
##       This should not need to be done if this is repeated on MARC1, however, at the time I did not have access/knowledge of this resource.

LUAD_CNV_dt <-data.table(LUAD_CNV_sorted)
complete_gene_list_dt <- data.table(complete_gene_list)
LUAD_CNV_dt$Num_Probes = NULL
write.csv(LUAD_CNV_dt, file = "LUAD_CNV_dt.csv")
complete_gene_list_dt <- complete_gene_list_dt[, c("uniprot_genename", "chromosome_name", "start_position", "end_position")]
write.csv(complete_gene_list_dt, file = "complete_gene_list_dt.csv")
setkey(LUAD_CNV_dt, Chromosome, Start, End)
LUAD_CNV_genenames <- foverlaps(complete_gene_list_dt, LUAD_CNV_dt, by.x = c("chromosome_name", "start_position", "end_position"), by.y = c("Chromosome", "Start", "End"),  maxgap=0, minoverlap=1, type = "any", mult = "all", nomatch=0, which=FALSE)
write.csv(LUAD_CNV_genenames, file = "LUAD_CNV_genenames.csv")



##################################
##
## DOWNLOADING THE CLINICAL DATA ASSOCIATED WITH THE LUAD CNV DATA 
##
##################################

#Create a GDC query to get clinical data from the Lung Adenocarcinoma dataset within the TCGA database.
queryLUAD_clinical <- GDCquery(project ="TCGA-LUAD",
                               data.category = "Clinical")

#Download the Lung Adeno clinical data.
LUAD_clinical <- tryCatch(GDCdownload(queryLUAD_clinical),
                          error=function(e) GDCdownload(queryLUAD_clinical, method = "client"))

#Access patient suvival data.
LUAD_clinical_patient <- GDCprepare_clinic(queryLUAD_clinical, clinical.info = "patient")
LUAD_clinical_patientsurvival <- LUAD_clinical_patient[,c("bcr_patient_barcode", "vital_status", "days_to_death")]

# Score patient survival data at 5 years (1825 days), 10 years (3650 days) and 15 years (5475 days) 
LUAD_clinical_patientsurvival <- LUAD_clinical_patientsurvival %>%
  mutate(Survival_Score5y=days_to_death/1825)
LUAD_clinical_patientsurvival$Survival_Score5y <- ifelse (LUAD_clinical_patientsurvival$vital_status == "Alive", 1, LUAD_clinical_patientsurvival$Survival_Score5y)

LUAD_clinical_patientsurvival <- LUAD_clinical_patientsurvival %>%
  mutate(Survival_Score10y=days_to_death/3650)
LUAD_clinical_patientsurvival$Survival_Score10y <- ifelse (LUAD_clinical_patientsurvival$vital_status == "Alive", 1 ,LUAD_clinical_patientsurvival$Survival_Score10y)

LUAD_clinical_patientsurvival <- LUAD_clinical_patientsurvival %>%
  mutate(Survival_Score15y=days_to_death/5475)
LUAD_clinical_patientsurvival$Survival_Score15y <- ifelse (LUAD_clinical_patientsurvival$vital_status == "Alive", 1 ,LUAD_clinical_patientsurvival$Survival_Score15y)

# Write files and print session information
write.csv(LUAD_clinical_patientsurvival, "LUAD_clinical_patientsurvival.csv")



##################################
##
## ANNOTATING CNV DATA WITH PATIENT DATA
##
##################################

# Splitting Sample ID into information about the patient - Patient ID and sample type.
LUAD_CNV_genenames_samplesplit <- within(LUAD_CNV_genenames, {
  Patient_ID <- substr(LUAD_CNV_genenames$Sample, 1, 12)
  Sample_Type_ID <- substr(LUAD_CNV_genenames$Sample, 14, 15)
})

#Reducing the data down into components required for next step (#### FOR GENENAMES ADD "uniprot_genename" AT END ####)
LUAD_CNV_genenames_samplesplit <- LUAD_CNV_genenames_samplesplit[, c("Patient_ID", "Sample_Type_ID", "Sample", "Segment_Mean")]

#Characterising the Sample Type ID into Sample Type (tumour vs normal vs control)
LUAD_CNV_genenames_samplesplit <- LUAD_CNV_genenames_samplesplit %>%
  mutate(Sample_Type=Sample_Type_ID)

for (i in c("01", "02", "03", "04", "05", "06", "07", "08", "09")){
  LUAD_CNV_genenames_samplesplit$Sample_Type <- ifelse (LUAD_CNV_genenames_samplesplit$Sample_Type == i, "Tumour", LUAD_CNV_genenames_samplesplit$Sample_Type)
}

for (j in c("10", "11", "12", "13", "14", "15", "16", "17", "18", "19")){
  LUAD_CNV_genenames_samplesplit$Sample_Type <- ifelse (LUAD_CNV_genenames_samplesplit$Sample_Type == j, "Normal", LUAD_CNV_genenames_samplesplit$Sample_Type)
}

for (k in c("20", "21", "22", "23", "24", "25", "26", "27", "28", "29")){
  LUAD_CNV_genenames_samplesplit$Sample_Type <- ifelse (LUAD_CNV_genenames_samplesplit$Sample_Type == k, "Control", LUAD_CNV_genenames_samplesplit$Sample_Type)
}

# Ensuring data is in the correct format and altering patient survival data table so the "Patient_ID" column is correctly labelled.
LUAD_clinical_patientsurvival <- data.table(LUAD_clinical_patientsurvival)
LUAD_CNV_genenames_samplesplit <- data.table(LUAD_CNV_genenames_samplesplit)
colnames(LUAD_clinical_patientsurvival)[1] <- "Patient_ID" 

# Merging the CNV data with the patient data based on the patient ID given by the TCGA database.
merge <- merge(LUAD_CNV_genenames_samplesplit, LUAD_clinical_patientsurvival, by="Patient_ID")
write.csv(merge, file = "LUAD_CNV_genenames_survival.csv")



##################################
##
## IMPORTING GENE LIST GENERATED BY PREVIOUS SCRIPTS
##
##################################

# Load lists of proteins with SH3 and PxxP regions 
PR_Class1and2 <- read.csv(file = "PR_Class1and2_containing_proteins_list.csv")
SH3_proteins <- read.csv(file = "SH3_domain_containing_proteins_list.csv")
PR_Class1and2$X <- NULL
SH3_proteins$X <- NULL
SH3_PR12_proteins <- rbind(PR_Class1and2, SH3_proteins)
SH3_PR12_proteins_dedup <- SH3_PR12_proteins[!duplicated(SH3_PR12_proteins$uniprot_genename),]  
SH3_PR12_proteins_dedup <- as.data.frame(SH3_PR12_proteins_dedup)

  

##################################
##
## FILTERING CNV DATA TO SHOW ONLY INFORMATION RELATING TO PROTEINS CONTAINING PXXP MOTIFS OR SH3 DOMAINS
##
##################################

# Filter CNV data to show only information on proteins containing canonical (Class 1 and 2) PxxP motifs or SH3 domains  
## NOTE: This step was performed on the MARC1 server details are below:
##                    DATE PERFORMED:     08DEC2017
##                    SUCCESSFUL JOB ID:  498770
##                    QNAME:              48core-3T.q
##                    TIME:               8.9 hours
##                    MEM:                9.5 MGBs
##                    MAXVMEME:           486 GB

LUAD_CNV_genenames <- read.csv(file = "LUAD_CNV_genenames.csv")
LUAD_CNV_genenames_reduced <- data.frame(LUAD_CNV_genenames$Sample, LUAD_CNV_genenames$Segment_Mean, LUAD_CNV_genenames$uniprot_genename)

LUAD_CNV_SH3_PR12 <- merge(LUAD_CNV_reduced, SH3_PR12, by.x="LUAD_CNV.uniprot_genename", by.y="uniprot_genename")
write.csv(LUAD_CNV_SH3_PR12, file="LUAD_CNV_SH3_PR12.csv")


##################################
##
## PRINCIPLE COMPONENT ANALYSIS
##
##################################

# Removing any duplicated and incomplete rows and reshaping the dataframe for PCA analysis
## NOTE: This step was performed on the MARC1 server details are below:
##                    DATE PERFORMED:     03JAN2018
##                    SUCCESSFUL JOB ID:  506468
##                    QNAME:              48core-3T.q
##                    TIME:               52.1 hours
##                    MEM:                17.5 MGBs
##                    MAXVMEME:           131 GB

LUAD_CNV <- read.csv(file = "LUAD_CNV_SH3_PR12.csv")

deduped_LUAD_CNV <- unique(LUAD_CNV)
LUAD_reshaped <- reshape(deduped_LUAD_CNV, idvar = "LUAD_CNV.Sample", timevar = "LUAD_CNV.uniprot_genename", direction = "wide")
write.csv(LUAD_reshaped, file="LUAD_reshaped.csv")

LUAD_reshaped_NAremoved <- LUAD_reshaped[complete.cases(LUAD_reshaped),]
write.csv(LUAD_reshaped_NAremoved, file="LUAD_reshaped_NAremoved.csv")

# The previous step produced an extra column for each gene during reshaping that was not required, this step removes these columns and renames the remaining columns to correctly reflect their contents. 
## NOTE: This step was performed on the MARC1 server details are below:
##                    DATE PERFORMED:     05JAN2018
##                    SUCCESSFUL JOB ID:  507575
##                    QNAME:              48core-3T.q
##                    TIME:               0.5 hours
##                    MEM:                32.1 KGBs
##                    MAXVMEME:           19 GB

###LUAD_reshaped <- read.csv(file = "LUAD_reshaped.csv")
LUAD_reshaped_redacted <- LUAD_reshaped[, -grep("X.", colnames(LUAD_reshaped))]
names(LUAD_reshaped_redacted) = gsub(pattern = "LUAD_CNV.Segment_Mean.", replacement = "", x = names(LUAD_reshaped_redacted))
write.csv(LUAD_reshaped_redacted, file = "LUAD_reshaped_redacted.csv")

## NOTE: This step was performed on the MARC1 server details are below - currently in process of performing.
# Removing any duplicated and incomplete rows for PCA analysis
LUAD_reshaped_redacted_complete <- LUAD_reshaped_redacted[complete.cases(LUAD_reshaped_redacted),]
write.csv(LUAD_reshaped_redacted_complete, file = "LUAD_reshaped_redacted_complete.csv")

# Standardising the Segment Mean for each gene
standardised_SM <- as.data.frame(scale(LUAD_reshaped_redacted_complete[, -1]))
#### Error in colMeans(x, na.rm = TRUE) : 'x' must be numeric
#### Calls: as.data.frame -> scale -> scale.default -> colMeans
#### Execution halted
standardised_SM <- as.matrix(standardised_SM)
write.csv(standardised_SM, file = "standardised_SM.csv")

# PCA
LUAD_pca <- prcomp(standardised_SM, center = TRUE, scale = TRUE)
write.csv(LUAD_pca, file = "LUAD_PCA.csv")
biplot_LUAD_PCA <- biplot(LUAD_pca)
dev.copy(biplot_LUAD_PCA, "biplot_LUAD_PCA.png")
dev.off()
summary_LUAD_PCA <- summary(LUAD_pca)
write.csv(summary_LUAD_PCA, file = "summary_LUAD_PCA.csv")



####### Multivariate analysis tool for CNV??? https://cran.r-project.org/web/packages/CNVassoc/CNVassoc.pdf


# Prints Session information for this code.
sessionInfo()